### DISCUSSION TITLE
dictionary , grammar checker

### DISCUSSION THREAD

**Discussion by vbatcnam** [vkbo/novelWriter#2142] [Help] [answered]
Hello, 
Can you tell me if it is possible to: 
- add a custom dictionary for invented words,
- integrate a grammar checker like "Grammalecte" (in french),
- add a plugin (or other) that analyzes sentences that are too long, repetitions and language tics?
If so, can you tell me how to do it?
And if not, what open source software do you use to do that?
I think It would be nice to be able to do all this with NovelWriter to avoid copying and pasting and duplicate text with different versions.

I also take this opportunity to say that including images (at least in places, objects and characters sections) would be very useful.
Currently I use Joplin or zettlr for my files but I sometimes have duplicates with the novelWriter files because I don't always think to synchronize them....
As for grammar correction, I use Zettlr, but the same: I have duplicates sometimes with different versions and switching from one to the other is quite tedious.
If you have a better method to tell me:)
(Forgive my English)

**Comment by vkbo** [comment, marked as answer]
> * add a custom dictionary for invented words,

Yes, there is a project word dictionary for this. You can find it in the Tools menu, and add words to it either directly or by right-clicking on an underlined word and select "Add to Dictionary".

> * integrate a grammar checker like "Grammalecte" (in french),
> * add a plugin (or other) that analyzes sentences that are too long, repetitions and language tics?

There are currently no text analysis tools in novelWriter. It's a pure writing app. Such features have been discussed and may be added at some point.

> And if not, what open source software do you use to do that?

I don't use such tools myself.

> I also take this opportunity to say that including images (at least in places, objects and characters sections) would be very useful. Currently I use Joplin or zettlr for my files but I sometimes have duplicates with the novelWriter files because I don't always think to synchronize them...

I get that people would like to also use novelWriter for research notes, but the plain text format is not well suited for rich notes. The note feature is meant for writing quick character profiles and such, much closer to the actual text. I keep my own research notes in Obsidian, and usually have it open next to novelWriter when writing.

There are still use cases for adding other content than text, and it's been discussed from time to time to allow adding attachments to notes, like images (like world maps and such). From next version it will be possible to add weblinks at least.

**Reply by rumplestilzken** [reply]
I run a LanguageTool server that i use with OnlyOffice and LibreOffice, i would be interested in taking a closer look at a plugin that would submit and analyze the results to such a server. They offer it as a free service, so you don't need to set up your own server, but I like my writing as far away from "the cloud" as possible, especially while drafting. 

I keep Obsidian vaults for all my work as well, they are submitting it a git repo i host using Gitea, along with submissions, important drafts or major revisions.

https://github.com/languagetool-org/languagetool
https://github.com/writingtool-org/writingtool

Thoughts?

**Reply by vkbo** [reply]
You can have a look at the closed ticket #515. The initial idea was to automate the parts needed to run a local server. But it is rather bulky software. On principle, I do not like to add features that involve sending data away from the local computer. I certainly do not want to host anything on behalf of the users. The security implications are too big if it is to be done properly.

General AI features are out of the question. See #2097. But adding text analysis features is interesting and absolutely something I'd like to look into at some point, but there is a long list of planned features, and I don't have a lot of capacity to add larger features.

There's also a plan to add a panel below the editor, possibly merged with the one in the viewer, to show various analysis info for the text. See #1345. I want to integrate the spell checker there. In general, I want to find a way to analyse text content that doesn't use the syntax highlighter to do it, like the spell checker currently does. Each text block in the editor has a meta data object associated with it where the spell check info is saved, but due to multi-threading limitations in how text is rendered in Qt, it hasn't been taken out of the syntax highlighter yet. I think solving this would be helpful also for text analysis.

**Comment by vbatcnam** [comment]
Great! Tanks.

**Comment by rumplestilzken** [comment]
In no way am i endorsing using AI for content creation. But using llm's for
style and grammar, spelling are great resources.

I'm getting in contact with you so I can see your vision. Setting up the
server, much less having the resources to host it locally is a nontrivial
process. But languagetool hosts it's API for free for people to use for a
similar effect to grammarly. And it didn't store the text, it just analyzes
it.

I will look over the tickets and get back with you at some point. I believe
i can help somewhere in the development and design of some of the planned
features.

**Reply by vkbo** [reply]
Please reply to the existing thread. Reply by email doesn't work properly. See #1393
